server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,314 @@
+            return [], []
+
+        self._logger.debug("========= starting matching algorithm =========")
+
+        searches = SortedList(searches, key=lambda s: s.average_rating)
+        possible_games = []
+        for index, search in enumerate(searches):
+
+            self._logger.debug("building game for %s", repr(search))
+
+            try:
+                participants = self.pick_neighboring_players(searches, index, team_size)
+                match = self.make_teams(participants, team_size)
+                game = self.assign_game_quality(match)
+                possible_games.append(game)
+            except NotEnoughPlayersException:
+                self._logger.warning("Couldn't pick enough players for a full game. Skipping this game...")
+            except UnevenTeamsException:
+                self._logger.warning("Failed to assign even teams. Skipping this game...")
+
+        if self._logger.isEnabledFor(logging.DEBUG):
+            self._logger.debug("got %i games", len(possible_games))
+            for game in possible_games:
+                self._logger.debug(
+                    "%s vs %s rating disparity: %i quality: %f",
+                    repr(game.match[0]),
+                    repr(game.match[1]),
+                    game.match[0].cumulated_rating - game.match[1].cumulated_rating,
+                    game.quality
+                )
+
+        matches = self.pick_noncolliding_games(possible_games)
+        for match in matches:
+            for team in match:
+                for search in team.get_original_searches():
+                    searches.remove(search)
+        return matches, list(searches)
+
+    def pick_neighboring_players(self, searches: List[Search], index: int, team_size: int) -> List[Search]:
+        """
+        Picks searches from the list starting with the search at the given index and then expanding in both directions
+        until there are enough players for a full game.
+
+        # Errors
+        May raise `NotEnoughPlayersException` if it can't find enough suitable searches to fill a game.
+        """
+        # We need to do this in two steps to ensure that index = 0 gives an empty iterator
+        lower = searches[:index]
+        lower = iter(lower[::-1])
+        higher = iter(searches[index:])
+        pick_lower = True
+        candidate = next(higher, None)
+        participants = [candidate]
+        number_of_players = len(candidate.players)
+
+        while number_of_players < team_size * 2:
+            candidate, prev = next(lower if pick_lower else higher, None), candidate
+            pick_lower = not pick_lower
+            if candidate is None:
+                if prev is None:
+                    raise NotEnoughPlayersException()
+                continue
+            if number_of_players + len(candidate.players) <= team_size * 2:
+                participants.append(candidate)
+                number_of_players += len(candidate.players)
```

Comment:
Of course this should happen in a way that doesn't break 5v5, etc. But I think it is ok to only add this optimization for 4v4.

Reply:
3v3 also has the edge case of (2,2,2). I am not against adding these two, but I am also not convinced what the benefit would be, because we catch these cases later and the matchmaker doesn't seem to be in a strong need for performance optimization.